### DISCUSSION TITLE
GitLab support for self hosted?

### DISCUSSION THREAD

**Discussion by ftzi** [lost-pixel/lost-pixel#372] [General]
Hey!

I am aware about these two:

https://github.com/lost-pixel/lost-pixel/discussions/180
https://github.com/lost-pixel/lost-pixel/issues/271

I am willing to develop my own GitLab solution and I would happily have it as a public repo.

I don't need much guidance, although some starting points would be very helpful.

Do you have some info you think could be useful for me to set it up?

Many thanks!

**Comment by d-ivashchuk** [comment]
Hey @SrBrahma! I see you are also in Vienna :D I think you could try using lost-pixel on GitLab as the open-source solution; I'm not sure if anything would prevent you from this. But for our SaaS platform, it would be quite impossible right now and would require some radical changes as we went all in for GitHub :D 

What are your expectations here, what do you want to achieve in the end?

**Reply by ftzi** [reply]
Ja! I have been living here for six months already! Great city!

I am a coworker of Christopher, who said he knows some of you (maybe including you?), at the same company where Dominik, the famous :smile:, previously worked!

I understand about the SaaS situation and it makes total sense.

What I minimally want to achieve is:

To have visual regression integrated on our self hosted CI, and on Storybook UI changes, it could open a Thread on the Merge Request (how GitLab calls PRs) with some useful and somehow reviewable content so it's required to be resolved by someone in order for the MR to be mergeable.

Not sure if that's the best solution but that's what I have in mind at the moment.

To open a Thread/Comment in the MR, it's just a matter of calling the GitLab API.

My point now is: How do I run Lost Pixel? Via CLI or .ts? What would be the output of it, if checking for diffs in Storybook? Does it create a directory with the snapshot diffs?

I think with that I can find my way through it.

Many thanks in advance!

**Reply by d-ivashchuk** [reply]
Ah you are working with Christopher, awesome! 

So lost pixel is a docker container so you would be able to run it easily on gitlab as well(my first guess here, I might be wrong). when lost pixel finishes it will exit with either 1 or 0 depending if something failed or not - this is the first clue that you could use. Also if you run lost pixel locally with for example `npx lost-pixel docker update` you will see that there are also files created that represent the current shots & difference shots that are essentially differences to your baselines.

You could tap into that from your ci as well, you could send those images anywhere yourself in some custom script and that would be it. Tell me if you need anything else here, would love to help!

**Reply by ftzi** [reply]
In a few days I will start working on it and I will return to you!

Thank you very much!

**Reply by ftzi** [reply]
@d-ivashchuk 

Hey!

I am a bit curious about the `npx lost-pixel docker update`. Does it always update the `baseline` dir content? If so, the `baseline` would match the content of `current` after it's run?

My current new idea of an initial MVP workflow is:

1) Dev does the changes.
2) If Lost Pixel detects diffs, the CI fails.
3) The `difference` and `current` can be seen from the CI Artifacts.
4) There will be a manual CI action to remove all the content in the `baseline` and copy the content of `current` there. A commit is made with these changes to the `baseline` dir. (Or maybe just run again with `update`?)
4.a) It can also be done locally by running a package.json script.
5) With that commit, another CI run would be done, now passing.

When I run it locally, as it is run by docker with root as user, I can't manually change/remove the contents of the baseline without sudo. This for example led to some situations where it messed git operations for me, like stashing/unstashing. I believe I could change my docker settings, but to avoid having other devs to change it as well, I came to this solution:

https://github.com/lost-pixel/lost-pixel/compare/main...SrBrahma:lost-pixel:main

**Reply by d-ivashchuk** [reply]
All together you are right about baseline update just moving files from current to baseline folder. I would still suggest to use that flow as it would be the most standard possible, but with simply running lost pixel not in update mode you could also decide on your own what to do next and how to handle failures and updates

**Reply by ftzi** [reply]
Thanks!

I will drop these changes on my fork; I can't run the `docker` in my CI as there is no docker command in my image, and running docker inside docker isn't a good thing as I found out. I managed to just run lost pixel directly without docker, by setting up playwright before!

Question: Is it possible to have the the screenshots to be taken using the total height of the page? By default it uses 720px. I know I can change the height to a specific px, but can it be 100% of the page content height?

**Reply by d-ivashchuk** [reply]
Maybe you can somehow hack it in beforeScreenshot but I highly doubt there is a good straightforward way of doing this 😀

**Reply by ftzi** [reply]
I will look into it! 🫡

**Reply by ftzi** [reply]
Hey, @d-ivashchuk!

So, apparently your project already has a code for it: https://github.com/lost-pixel/lost-pixel/blob/3ac2e30343c259397f211cee9747695a86500c33/src/shots/shots.ts#L101C1-L126C4 (although maybe the fullScreenMode boolean is inverted there?)

The issue in my project apparently is: In cartesian stories, where the content can be super tall, due to whatever reason the div height is only tall on a child. The parent height is not that tall.

<img width="1102" alt="Screenshot 2024-02-21 224727" src="https://github.com/lost-pixel/lost-pixel/assets/28177550/e56d96d3-3547-40fb-9555-496e4894886e">

<img width="980" alt="Screenshot 2024-02-21 224738" src="https://github.com/lost-pixel/lost-pixel/assets/28177550/15cb456f-92b6-478f-9188-61ffa7ce1ff7">

Maybe this code part could allow custom height query extractors, as I don't know if I can tweak the cartesian story behavior to not shrink the parent height.

https://github.com/lost-pixel/lost-pixel/blob/3ac2e30343c259397f211cee9747695a86500c33/src/shots/utils.ts#L108C1-L114C11

I will investigate more later.

**Reply by ftzi** [reply]
Also, does it currently have a way to retry a page that failed? It would be useful for flaky pages. As it doesn't take long to take a single screenshot and to compare, it would be very useful if it could retry ~3 times and log the flaky ones

**Reply by d-ivashchuk** [reply]
https://github.com/lost-pixel/lost-pixel/blob/ab0704a37825429a77d3829af1f1e44ff79c0286/src/config.ts#L317-L321

not really well documented but we have it of course :D

**Reply by ftzi** [reply]
Ohhh! This is awesome! I will PR it for the docs! Many thanks!

**Reply by ftzi** [reply]
@d-ivashchuk 

We have it now in our CI! Thank you so much for your support! Besides the ci file, I made a .ts wrapper around lost-pixel to change the output a bit, to support the workflow i designed. I will later make a repo for it; maybe we can think of making a PR here. Also, of course, I will make a doc page on how to integrate it with GitLab ;)

It is still in an early stage of a bit of flakiness; letting the team know about it and making it more stable.

I've been thinking about the retries. Currently, with the flakynessRetries, it takes the first screenshot and N <= Retries screenshots until the same hash is obtained. With the flakynessRetries >= 1, the total number of screenshots is at least 2.

If, in the comparison, it fails, there isn't a further retry.

So, my situation is that a single retry happens, and the same hash is obtained, but it's a false positive. I apparently can only increase the times, which applies to all the screenshots.

The best, I believe, would be that it wouldn't wait for all the screenshots to happen to compare them all (which is super fast); it would take a screenshot, compare, and retry if needed. The total number of screenshots would be at least 1 if it succeeded at first.

This would not be an easy change in the code, however.

**Reply by d-ivashchuk** [reply]
Yeah that would be pretty interesting and complex change! I would love to see your gitlab integration 😄

**Reply by ftzi** [reply]
Handled it here! https://github.com/lost-pixel/lost-pixel/pull/385
